File: problems/0037_sudoku_solver/0037_sudoku_solver.java
class Solution {
    boolean[][] rows = new boolean[9][10];
    boolean[][] cols = new boolean[9][10];
    boolean[][] blocks = new boolean[9][10];

    public void solveSudoku(char[][] board) {
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (board[r][c] != '.') {
                    int n = board[r][c] - '0';
                    int blockIdx = getBlockIdx(r, c);
                    rows[r][n] = true;
                    cols[c][n] = true;
                    blocks[blockIdx][n] = true;
                }
            }
        }
        backtrack(board, 0, 0);
    }

    private boolean backtrack(char[][] board, int r, int c) {
        if (r == 9)
            return true;
        if (c == 9)
            return backtrack(board, r + 1, 0);
        if (board[r][c] != '.')
            return backtrack(board, r, c + 1);

        int blockIdx = getBlockIdx(r, c);
        for (int n = 1; n < 10; n++) {
            if (!rows[r][n] && !cols[c][n] && !blocks[blockIdx][n]) {
                rows[r][n] = cols[c][n] = blocks[blockIdx][n] = true;
                board[r][c] = (char) ('0' + n);

                if (backtrack(board, r, c + 1))
                    return true;

                board[r][c] = '.';
                rows[r][n] = cols[c][n] = blocks[blockIdx][n] = false;
            }
        }
        return false;
    }

    private int getBlockIdx(int r, int c) {
        return (r / 3) * 3 + c / 3;
    }
}

File: problems/0037_sudoku_solver/0037_sudoku_solver.py
class Solution:
    def __init__(self):
        self.rows = [[False] * 10 for _ in range(9)]
        self.cols = [[False] * 10 for _ in range(9)]
        self.blocks = [[False] * 10 for _ in range(9)]

    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        for r in range(9):
            for c in range(9):
                if board[r][c] != '.':
                    n = int(board[r][c])
                    block_idx = self.getBlockIdx(r, c)
                    self.rows[r][n] = True
                    self.cols[c][n] = True
                    self.blocks[block_idx][n] = True
        self.backtrack(board, 0, 0)
    
    def backtrack(self, board: List[List[str]], r : int, c :int) -> bool:
        if r == 9:
            return True
        if c == 9:
            return self.backtrack(board, r + 1, 0)
        if board[r][c] != '.':
            return self.backtrack(board, r, c + 1)
        block_idx = self.getBlockIdx(r, c)
        for n in range(1, 10):
            if not self.rows[r][n] and not self.cols[c][n] and not self.blocks[block_idx][n]:
                self.rows[r][n] = self.cols[c][n] = self.blocks[block_idx][n] = True
                board[r][c] = str(n)
                if self.backtrack(board, r, c + 1):
                    return True
                
                ## Rollback
                board[r][c] = '.'
                self.rows[r][n] = self.cols[c][n] = self.blocks[block_idx][n] = False

        return False

    def getBlockIdx(self, r: int, c: int) -> int:
        return (r // 3) * 3 + (c // 3)